### PR TITLE
[feat] engine: implementation of yandex (web, images)

### LINKED ISSUE #961: Yandex engine request
**Working URL to the engine**

https://yandex.ru/
https://yandex.eu/

(and more country specific TLD)

**Why do you want to add this engine?**

Yandex is the only big company backed search engine that isn't owned by a western company. 

It's the only one that doesn't filter anything (yes, this includes things about the current war) - Edit 2024-06-04: at least it hasn't be proven to be filtering stuff.

It's also only the only engine that doesn't "biases" its results for the sake of being politically correct/inclusive. (Example, I heard about recently: [DDG](https://duckduckgo.com/?q=blonde+braids&iax=images&ia=images) / [Yandex](https://yandex.eu/images/search?text=blonde%20braids)) - Edit 2024-06-04: some testing was done, Yandex.eu is now biased (though a lot less than DDG and western engines), the "problem" isn't present on Yandex.ru ([see here](https://yandex.ru/images/search?text=blonde%20braids)) (so both will have to be added)

**Features of this engine**

See previous points.

**How can SearXNG fetch the information from this engine?**

General, Images,Maps, Videos

**Additional context**

Support for it was removed in #709 (but it basically was broken at this point)

### PR DESCRIPTION
**Pull request by Austin-Olacsi**
## What does this PR do?

adds yandex engine for web and image results.

## Why is this change important?

it's about time we have yandex in searxng again.

## How to test this PR locally?

it's set to inactive in settings.yml because of captcha. you need to remove that from the settings.yml.
then you can use: make run

## Author's checklist

my old draft: https://github.com/searxng/searxng/pull/3201
engine request: https://github.com/searxng/searxng/issues/961

## Related issues
Closes #961

### CODE REVIEW

File: searx/engines/yandex.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    if params['pageno'] > 1:
+        query_params_web.update({"p": params["pageno"] - 1})
+        query_params_images.update({"p": params["pageno"] - 1})
+
+    params["cookies"] = {'cookie': "yp=1716337604.sp.family%3A0#1685406411.szm.1:1920x1080:1920x999"}
+
+    if search_type == 'web':
+        params['url'] = f"{base_url_web}?{urlencode(query_params_web)}"
+    elif search_type == 'images':
+        params['url'] = f"{base_url_images}?tmpl_version=releases&{urlencode(query_params_images)}"
```
Comment (Bnyro) [inline searx/engines/yandex.py]: `tmpl_version` should likely be part of `query_params_images` as well to keep things cleaner

File: searx/engines/yandex.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    return params
+
+
+def response(resp):  # pylint: disable=too-many-locals
+    if search_type == 'web':
+        if (resp.url).path.startswith('/showcaptcha'):
```
Comment (Bnyro) [inline searx/engines/yandex.py]: repeated in line 85, should be outside the if condition.

the brackets around `resp.url` are not needed either here

File: searx/engines/yandex.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            raise SearxEngineCaptchaException()
+
+        html_data = html.fromstring(resp.text)
+        html_sample = unescape(html.tostring(html_data, encoding='unicode'))
+
+        start_tag = '{"location":"/images/search/'
+        end_tag = 'advRsyaSearchColumn":null}}'
+
+        start_index = html_sample.find(start_tag)
+        start_index = start_index if start_index != -1 else -1
```
Comment (Bnyro) [inline searx/engines/yandex.py]: what it does is `start_index = start_index`, if that's the intention the line can be dropped

File: searx/engines/yandex.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        if resp.url.path.startswith('/showcaptcha'):
+            raise SearxEngineCaptchaException()
+
+        html_data = html.fromstring(resp.text)
+        html_sample = unescape(html.tostring(html_data, encoding='unicode'))
+
+        start_tag = '{"location":"/images/search/'
+        end_tag = 'advRsyaSearchColumn":null}}'
+
+        start_index = html_sample.find(start_tag)
```
Comment (Bnyro) [inline searx/engines/yandex.py]: This could be simplified using https://github.com/searxng/searxng/blob/master/searx/utils.py#L356

**Review by return42 — APPROVED** [review]
Thank you for devoting yourself to the topic again. Even if we can't prevent the CAPTCHA yet, I think it's OK if we add the engines as "inactive".  Maybe someone will see this code and find a solution to avoid the CAPTCHA.

I squashed your two commits into one and fixed a nit / [diff](
https://github.com/searxng/searxng/compare/6014340a1e5b4a7714ea27430f39f289b07ab0b4..e04ff2c8d0f3d8f2f2720e95c6d31e31956af91f)

**Comment by ikkqqstudio** [comment]
Any updates in 2025? And Why is it taking so longer than 4 months to add Yandex into The open-source search engine?

**Comment by return42** [comment]
> And Why is it taking so longer than 4 months to add Yandex into The open-source search engine?

It is in .. but

> it's set to inactive in settings.yml because of captcha.

Maintainers can activate the engine in their settings.

@ikkqqstudio Once you show us how to avoid the CAPTCHA, we can enable Yandex by default ... We are just waiting for your cooperation ;-)

**Comment by ikkqqstudio** [comment]
There is a Yandex search API, but it requires Paid Cloud subscription

**Pull request merged**